Improve error handling with `UResult`

tertsdiepraam opened the pull request:
Closes https://github.com/uutils/coreutils/issues/2336.

The full rationale behind this PR is explained in that issue.

# Basics
With this PR, a typical `uumain` signature would become something like:
```rust
#[uucore_procs::gen_uumain]
fn uumain(args: impl uucore::Args) -> UResult<()>
```

Where `UResult` is roughly:
```rust
pub type UResult<T> = Result<T, UError>;

enum UError {
    Common(UCommonError),
    Custom(Box<dyn UCustomError),
}
```

`UCommonError` is an enum containing errors that appear across multiple utils and provides default messages and exit codes. `UCustomError` is a trait that extends `Error` and requires a `code` method that returns the exit code for the error. The `Display` trait (required by `Error`) is used for printing the error message.

# Mapping Common Errors to Custom Errors
Additionally, `UResult` implements 2 methods: `map_custom_code` and `map_custom_message_and_code` which take a function that maps from a `UCommonError` to `i32` and `(i32, String)` respectively, turning `UCommonError` into `UCustomError`, which custom exit codes and messages.

> Note: Maybe those mapper functions should return `Option<i32>` and `Option<(i32, String)>`, so that not all `UCommonError` variants need to be taken into account?

# ExitCode
There is an `ExitCode` type that implements `UCustomError` and takes an `i32`. Most utils should be easy to convert to this model, by changing the returned `i32`'s with `ExitCode`. For example:
```rust
return 1;
// becomes
return Err(ExitCode(1).into());
```

This is not very idiomatic, but I think it works for the transition period.

# `set_exit_code` & `show!`
Finally, there is a `set_exit_code` function, which sets the exit code for the case where `Ok(())` is returned. This is useful for non-fatal errors, where we do want to affect the exit code, but do not want to stop the program.

`set_exit_code` works nicely with the `show!` macro, which is essentially `show_error!`, but it takes a `UError`, `UCommonError` or `UCustomError` and prints the error message.

# Next steps
Once this is merged, we can start porting the other utils. After that, we can remove the proc macro attribute and put that logic into `uucore_procs::main`.

---

I have applied all this to `ls` and intent to do a few more before marking this ready for merging. Meanwhile, I would appreciate any feedback.

tertsdiepraam commented:
@miDeb I hope this matches what you had in mind!

miDeb commented on src/uu/ls/src/ls.rs:
```diff
@@ -125,6 +126,32 @@ pub mod options {
+impl Display for LsError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            LsError::InvalidLineWidth(s) => write!(f, "invalid line width: ‘{}’", s),
```
This should use `'` quotes as well afaict.

miDeb commented on src/uucore/src/lib/mods/error.rs:
```diff
@@ -0,0 +1,198 @@
+    }
+}
+
+/// A common error with an overriden exit code. This exit code is returned
```
nit:
```suggestion
/// A common error with an overridden exit code. This exit code is returned
```

miDeb commented on src/uucore/src/lib/macros.rs:
```diff
@@ -21,6 +21,14 @@ macro_rules! executable(
+macro_rules! show(
+    ($err:expr) => ({
+        uucore::error::set_exit_code($err.code());
+        eprintln!("{}: {}", executable!(), $err);
```
I think the use of `executable!()` means that this macro can only be used in individual utilities, but not in `uucore`, because `executable!()` looks at the path of the current file and guesses the utility name from that.
Maybe we can instead change how `executable!()` works and make it return the actual command that was executed?

tertsdiepraam replied:
That's true. In what kind of situations would you like to use it in `uucore`? Making a more robust implementation won't be easy, because it needs to work both for the individual binaries and the `coreutils` binary.

miDeb replied:
I didn't have a particular use case in mind, so I guess we can leave it as it is.

miDeb commented on src/uucore/src/lib/mods/error.rs:
```diff
@@ -0,0 +1,198 @@
+
+/// Set the exit code for the program if `uumain` returns `Ok(())`.
+pub fn set_exit_code(code: i32) {
+    EXIT_CODE.store(code, Ordering::SeqCst);
```
Do we want to add some logic here to make sure we don't call it multiple times with different exit codes? Maybe it's a clear indication of a bug if we first store one exit code and then another, but I don't know.

tertsdiepraam replied:
Maybe, but I think it's too early to say. I would have to look at some more utils to see how this might be used.

miDeb commented on src/uucore/src/lib/mods/error.rs:
```diff
@@ -0,0 +1,198 @@
+
+impl Display for ExitCode {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
+        write!(f, "")
```
I think we can just return Ok instead.
```suggestion
        Ok(())
```

tertsdiepraam replied:
Ah, of course! Thanks!

miDeb commented:
Thank you for implementing this! I hope what I had in mind makes sense for you as well.
> Note: Maybe those mapper functions should return Option<i32> and Option<(i32, String)>, so that not all UCommonError variants need to be taken into account?
Sounds like a good idea! For even more flexibility the return type could be `(Option<i32>, Option<String>)`, but I don't know if that's needed.

tertsdiepraam commented:
> I hope what I had in mind makes sense for you as well.

Definitely! Thanks for your comments!

tertsdiepraam commented:
I added a `context` method on `std::io::{Error, Result}` that convert into a `UCommonError` or `UResult` respectively via two traits `ToUIoError` and `ToUIoResult` (other suggestions for the naming of any of these additions are welcome).

I tried to do it an intermediate `into` to avoid the traits, but then type annotations were needed and the code got cumbersome, so I went with this easier option. I also converted `mkdir` where some of the advantages of this whole thing finally become clear, because I was able to make some functions return `UResult` and use `?` a few times.

One thing I'm not sure about: should `show!` set the exit code? It does right now and it makes the code quite clean, but also very implicit, so I'm not sure whether it's the right call. Maybe the name needs to be more explicit? Like `show_and_exit_code`? On the other hand, I think there's rarely a situation where you wouldn't want to set the exit code.

tertsdiepraam commented:
I've now used it for `ls`, `touch`, `mkdir` and `mktemp`, with slightly different approaches:
- `ls`: showing errors directly without returning them at all, only 1 custom error, was very easy to convert
- `touch`: very similar to `ls`
- `mkdir`: I was able to make some functions return `UResult`, but the util always returns `Ok` uses the exit code set by `set_exit_code`. I also made another macro for this one called `show_if_err`, which prints the message of a `UResult` if there is one.
- `mktemp`: I changed this one a lot, the addition of `UResult` lead to (in my opinion) more natural organization of the code. I was able to use the `io::Error` wrappers and make an `enum` with custom errors.

I think this approach really shines in utils with a lot of fatal errors. In utils which don't it is a little bit cumbersome, although it also feels more robust any possibly easier to work with once the code is more adapted to it.

Next steps after this:
- Porting more utils
- Implementing custom clap output to match GNU (I'm thinking of putting this behind a feature flag, but we can discuss that in another issue)

All in all, I think that this PR now provides a solid foundation which we could still tweak later if necessary, so I'm putting this up for review.

tertsdiepraam commented:
@miDeb Would you mind reviewing this?

miDeb commented on src/uu/ls/src/ls.rs:
```diff
@@ -38,8 +39,8 @@ use std::{
 };
-
 use term_grid::{Cell, Direction, Filling, Grid, GridOptions};
+use uucore::error::{set_exit_code, FromIo, UCustomError, UResult};
```
I wonder if we could collect this is a `uucore::error::prelude` so that we can just import that prelude in all utilities (`use uucore::error::prelude::*;`), instead of importing everything separately.
Or maybe we can just do `use uucore::error::*`.

miDeb reviewed (approved):
Looks really good! :+1:

sylvestre commented:
Sorry but could you please rebase the patch?

sylvestre commented:
Seems that it regressed a few GNU tests:
`GNU testsuite = TOTAL: 601 / PASS:  160 / FAIL:  328 / ERROR: 23`
currently:
`GNU testsuite = TOTAL: 601 / PASS:  165 / FAIL:  324 / ERROR: 26`

Could you please have a look?

tertsdiepraam commented:
> Sorry but could you please rebase the patch?

Done!

>  regressed a few GNU tests

I hope that this was just due to it being outdated. I'll check when the CI has run the GNU tests.

tertsdiepraam commented:
GNU is now at:
```
GNU testsuite = TOTAL: 601 / PASS:  167 / FAIL:  319 / ERROR: 29
```
So it seems all good now!

sylvestre commented:
I like the change. @rivy @Arcterus wdyt?

rivy commented:
I like it! I think the examples created here by @tertsdiepraam show that it can lead to better and more uniform util code.

At least for me, it's a complex chunk to grasp (but then again proc_macros have always been brain-benders for me). I do think it needs some iteration to polish the semantics and nomenclature and some internal developer documentation (rationale/need and usage) at least to help others like me to "grok" the implementation and use.

sylvestre commented:
> At least for me, it's a complex chunk to grasp (but then again proc_macros have always been brain-benders for me). I do think it needs some iteration to polish the semantics and nomenclature and some internal developer documentation (rationale/need and usage) at least to help others like me to "grok" the implementation and use.

Because of the size of the patch, what do you think about landing it this way and polish it in a new PR?

tertsdiepraam commented:
@rivy 

>  it's a complex chunk to grasp (but then again proc_macros have always been brain-benders for me).

For what it's worth: the macro takes the function it's given, renames it to `uumain_result` (which is actually not necessary, I'll remove that soon), executes the function and returns the exit code and prints a error message. This is what the result looks like:

```rust
pub fn uumain(args: impl uucore::Args) -> i32 {
    fn uumain_result(args: impl uucore::Args) -> UResult<()> {
        // content of the original uumain
    }
    let result = uumain_result(args);
    match result {
        Ok(()) => uucore::error::get_exit_code(),
        Err(e) => {
            let s = format!("{}", e);
            if s != "" {
                show_error!("{}", s);
            }
            e.code()
        }
    }
}
```

> I do think it needs some iteration to polish the semantics and nomenclature

Agreed, I wasn't always sure what to name the types. If you have any suggestions, please let me know!

> internal developer documentation (rationale/need and usage)

I could add some more documentation in the form of `rustdoc` comments in the file with the rationale either in this PR or another. I already tried to write some docs for `UIoError` would you be happy with similar docs on the other public types as well?

rivy commented:
> 
> 
> > At least for me, it's a complex chunk to grasp (but then again proc_macros have always been brain-benders for me). I do think it needs some iteration to polish the semantics and nomenclature and some internal developer documentation (rationale/need and usage) at least to help others like me to "grok" the implementation and use.
> 
> Because of the size of the patch, what do you think about landing it this way and polish it in a new PR?

👍🏻 

We can use the merged code to iterate on both the code and it's use in a few of the utils (with plan to later push to the rest of the utils).

Before a merge though, @tertsdiepraam , would you mind squashing the commits into descriptive, logically separate pieces?

Something like:

```
uucore: (error) add more standardized and idiomatic error handling (adds UError/UResult)
uucore_procs: add (temporary?) `gen_uumain` proc macro attribute supporting new error handling
touch: refactor for new error handling + ...
ls: ...
mktemp: ...
tests: adapt 'mktemp' tests to new error handling
tests: fix 'group' tests for ...
```

It helps with review/understanding and later code revisions. I'm happy to help if you need it.

rivy commented:
> For what it's worth: the macro takes the function it's given, renames it to `uumain_result` (which is actually not necessary, I'll remove that soon), executes the function and returns the exit code and prints a error message. This is what the result looks like:

Yes, I see. I've been playing with it a bit to see if I could figure out an adapter so that it could be moved directly into the `main!()` macro and remove some of the required util code changes. I'm not sure if it's possible yet...

> I could add some more documentation in the form of `rustdoc` comments in the file with the rationale either in this PR or another. I already tried to write some docs for `UIoError` would you be happy with similar docs on the other public types as well?

👍🏻 

I think a little more exposition within `rustdoc` comments at the appropriate types would be great.

tertsdiepraam commented:
For the macro, I thought of the attribute as an temporary thing, which we can remove when all the utils use the result type

tertsdiepraam commented:
@rivy I've added some more documentation, let me know if anything is still unclear!

I've also remembered why I implemented this with an attribute instead of putting the logic in `main!`. Putting it in `main!` worked for the individual utils, but was hard to implement for the busybox-style binary, because `uumain` is imported directly, so `main!` is not used. I think renaming the macro would be good though, maybe `#[uucore::main]`? That would be similar to `#[tokio::main]`. Or `#[uucore::uresult_main]`?

sylvestre commented:
sorry but we have a few conflicts now, could you please fix them?

tertsdiepraam commented:
@rivy @sylvestre I cleaned up the git history, fixed the conflicts, added documentation and the tests pass (except for an unrelated `cp` test failure)!
